### ISSUE TITLE
Persistent connections

### ISSUE THREAD

**Issue by zeljkobekcic** [geopy/geopy#363]
It would be very nice  to use something like `requests.Session` to keep connections persistent hence to geocode faster multiple objects. 

Is it possible to add this?

**Comment by KostyaEsmukov** [comment]
That's something I'm working on for the next major release geopy 2.

The main problem is keeping support for arbitrary `ssl_context` instances. It's certainly possible with a number of hacks (I already got a working version), but I need some time to polish it a little.

So, yeah, it is absolutely possible and it's going to be added soon.

**Comment by zeljkobekcic** [comment]
Can I help you with that somehow?

**Comment by KostyaEsmukov** [comment]
I hope to shape it up by the end of next week. After that, it would be great if you could test it and give some feedback. Maybe then it will be possible to split up some work.

I'll let you know in this thread when it will be pushed. Thank you for the offer, I appreciate it!

**Comment by KostyaEsmukov** [comment]
I have pushed a somewhat preliminary implementation of this feature to the `2` branch: https://github.com/geopy/geopy/pull/335

Assuming that `requests` is installed, it should just work, i.e. the requests adapter would be used automatically.

I already see the improvement in the test suite running times: with urllib it took ~200 seconds on my machine, but with requests it takes ~130 seconds, which is certainly great news!

@zeljkobekcic would you be able to play around with this and give some feedback? Thanks in advance!

The `2` branch can be installed with pip:

```
pip uninstall geopy
pip install 'https://github.com/geopy/geopy/archive/2.zip#egg=geopy[requests]'
```

**Comment by zeljkobekcic** [comment]
Nice, will play with it around and provide you feedback. I should be able to test it by the end of this or the next week.

**Comment by justahuman1** [comment]
I'll test it out as well. I am loving this library though, thank you.

**Comment by zeljkobekcic** [comment]
@KostyaEsmukov Looks stable to me. Have had no problems with it

**Comment by KostyaEsmukov** [comment]
@zeljkobekcic That's great, thank you! Did you notice any performance improvement?

**Comment by zeljkobekcic** [comment]
Of course. I think it was from about 20s for 100 requests down to 13s for 100 request. Will verify this the next week.

Another improvement would be something like a method `geocode_mutliple` which does asynchronous requests with an iterable of queries. This would definetly improve the speed a lot!

**Comment by KostyaEsmukov** [comment]
> Another improvement would be something like a method geocode_mutliple which does asynchronous requests with an iterable of queries. This would definetly improve the speed a lot!

There's no need to make that a part of the Geocoder interface. One could simply use threading for that.

Some geocoder services do support batch geocoding via API (see #223), but most of them -- don't. So making such method a part of the interface would be confusing, because the geocoder services which don't provide batch geocoding methods would ratelimit requests much more often than those who do.

Although I guess that a threaded apply would have been a good new `extras` module (along with [RateLimiter](https://geopy.readthedocs.io/en/1.20.0/#geopy.extra.rate_limiter.RateLimiter)). However, a special care must be taken around threadsafety, as requests's Session object is, for one, not guaranteed to be threadsafe (see https://github.com/not-kennethreitz/requests/issues/2766).

Also, I guess, the asyncio interface, which I'm working on for v2, would make batch geocoding even more convenient.

**Comment by KostyaEsmukov** [comment]
Merged to master #417 (which adds support for `requests`) and #419 (which adds an optional asyncio support). They will be released as geopy 2.0 on pypi in the coming days.
